services/121-service/src/payments/fsp-integration/safaricom/safaricom.service.ts

```diff
@@ -173,104 +111,77 @@ export class SafaricomService
+        ? safaricomCustomData.requestResult['OriginatorConversationID']
+        : 'Invalid Request';
+
+    safaricomTransferEntity.transactionId = transaction.id;
+
+    await this.safaricomTransferRepository.save(safaricomTransferEntity);
   }
 
-  public async processSafaricomResult(
-    safaricomPaymentResultData: any,
-    attempt = 1,
+  public async processSafaricomCallback(
+    safaricomPaymentResultData: SafaricomTransferCallbackJobDto,
+    _attempt = 1,
   ): Promise<void> {
-    const safaricomDbRequest = await this.safaricomRequestRepository
-      .createQueryBuilder('safaricom_request')
-      .leftJoinAndSelect('safaricom_request.transaction', 'transaction')
+    const job = await this.safaricomCallbackQueue.add(
+      ProcessNamePayment.callbackPayment,
+      safaricomPaymentResultData,
+    );
+    await this.redisClient.sadd(getRedisSetName(job.data.programId), job.id);
+  }
+
+  public async getSafaricomTransferByOriginatorConversationId(
+    originatorConversationId: string,
+  ): Promise<SafaricomTransferEntity[]> {
+    const safaricomTransfers = await this.safaricomTransferRepository
+      .createQueryBuilder('safaricom_transfer')
+      .leftJoinAndSelect('safaricom_transfer.transaction', 'transaction')
       .where(
-        'safaricom_request.originatorConversationID = :originatorConversationID',
+        'safaricom_transfer."originatorConversationId" = :originatorConversationId',
         {
-          originatorConversationID:
-            safaricomPaymentResultData.Result.OriginatorConversationID,
+          originatorConversationId: originatorConversationId,
         },
       )
       .getMany();
-    if (safaricomDbRequest[0] === undefined && attempt <= 3) {
-      attempt++;
-      await waitFor(850);
-      await this.processSafaricomResult(safaricomPaymentResultData, attempt);
-      return;
-    }
-
-    let paymentStatus: StatusEnum | null = null;
-
-    if (
-      safaricomPaymentResultData &&
-      safaricomPaymentResultData.Result &&
-      safaricomPaymentResultData.Result.ResultCode === 0
-    ) {
-      paymentStatus = StatusEnum.success;
-    } else {
-      paymentStatus = StatusEnum.error;
-      safaricomDbRequest[0].transaction.errorMessage =
-        safaricomPaymentResultData.Result.ResultDesc;
-    }
-
-    safaricomDbRequest[0].status = paymentStatus;
-    safaricomDbRequest[0].paymentResult = safaricomPaymentResultData;
-
-    const safaricomCustomData = {
-      ...safaricomDbRequest[0].transaction.customData,
-    };
-    safaricomCustomData['paymentResult'] = safaricomPaymentResultData;
-    safaricomDbRequest[0].transaction.status = paymentStatus;
-    safaricomDbRequest[0].transaction.customData = safaricomCustomData;
+    return safaricomTransfers;
+  }
 
-    await this.safaricomRequestRepository.save(safaricomDbRequest);
-    await this.transactionRepository.save(safaricomDbRequest[0].transaction);
+  public async updateSafaricomTransfer(
```

Comment:
What is the added value of putting this in a function? If needed, it should go into the repo.

Reply:
processed